From 67432da7c8629bc509e80e738972234169b69c92 Mon Sep 17 00:00:00 2001
From: Harsh Kumar Bijlani <hbijlani@qti.qualcomm.com>
Date: Mon, 26 May 2025 11:48:06 +0530
Subject: [PATCH] UPSTREAM: CP: wifi: ath12k: Free ml_peer_id in failure cases

Currently the logic for freeing ml_peer_id is missing in failure
scenarios during station state transitions.

This can result in the assocation failure for new stations when
there is no free bit in ml_peer_id bitmap to be given for new
station.

In order to address this, add code to free the bit for ml_peer_id
in the bitmap in failure scenarios during station state transitions.

Signed-off-by: Harsh Kumar Bijlani <hbijlani@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/mac.c  | 17 ++++++-----------
 drivers/net/wireless/ath/ath12k/peer.c | 12 ++++++++++++
 drivers/net/wireless/ath/ath12k/peer.h |  1 +
 3 files changed, 19 insertions(+), 11 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 2a7e982..5c0d230 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -8267,7 +8267,6 @@ static int ath12k_mac_assign_link_sta(struct ath12k_hw *ah,
 static void ath12k_mac_ml_station_remove(struct ath12k_vif *ahvif,
 					 struct ath12k_sta *ahsta)
 {
-	struct ieee80211_sta *sta = ath12k_ahsta_to_sta(ahsta);
 	struct ath12k_hw *ah = ahvif->ah;
 	struct ath12k_link_vif *arvif;
 	struct ath12k_link_sta *arsta;
@@ -8296,10 +8295,7 @@ static void ath12k_mac_ml_station_remove(struct ath12k_vif *ahvif,
 		ath12k_mac_free_unassign_link_sta(ah, ahsta, link_id);
 	}
 
-	if (sta->mlo) {
-		clear_bit(ahsta->ml_peer_id, ah->free_ml_peer_id_map);
-		ahsta->ml_peer_id = ATH12K_MLO_PEER_ID_INVALID;
-	}
+	ath12k_peer_ml_free(ah, ahsta);
 }
 
 static int ath12k_mac_handle_link_sta_state(struct ieee80211_hw *hw,
@@ -8487,13 +8483,9 @@ int ath12k_mac_op_sta_state(struct ieee80211_hw *hw,
 			ath12k_hw_warn(ah, "unable to create ath12k_dp_peer for sta %pM",
 				       sta->addr);
 
-			if (sta->mlo) {
-				clear_bit(ahsta->ml_peer_id, ah->free_ml_peer_id_map);
-				ahsta->ml_peer_id = ATH12K_MLO_PEER_ID_INVALID;
-			}
 			ah->num_ml_peers--;
 
-			goto exit;
+			goto ml_peer_id_free;
 		}
 
 		ret = ath12k_mac_assign_link_sta(ah, ahsta, arsta, ahvif,
@@ -8501,7 +8493,7 @@ int ath12k_mac_op_sta_state(struct ieee80211_hw *hw,
 		if (ret) {
 			ath12k_hw_warn(ah, "unable assign link %d for sta %pM",
 				       link_id, sta->addr);
-			goto exit;
+			goto peer_delete;
 		}
 
 		/* above arsta will get memset, hence do this after assign
@@ -8586,6 +8578,9 @@ int ath12k_mac_op_sta_state(struct ieee80211_hw *hw,
 peer_delete:
 	if (ret)
 		ath12k_dp_peer_delete(&ah->dp_hw, sta->addr);
+ml_peer_id_free:
+	if (ret)
+		ath12k_peer_ml_free(ah, ahsta);
 exit:
 
 	if (ret && is_recovery)
diff --git a/drivers/net/wireless/ath/ath12k/peer.c b/drivers/net/wireless/ath/ath12k/peer.c
index 105e985..1a85f93 100644
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -330,6 +330,18 @@ u16 ath12k_peer_ml_alloc(struct ath12k_hw *ah)
 	return ml_peer_id;
 }
 
+void ath12k_peer_ml_free(struct ath12k_hw *ah, struct ath12k_sta *ahsta)
+{
+	struct ieee80211_sta *sta = ath12k_ahsta_to_sta(ahsta);
+
+	lockdep_assert_wiphy(ah->hw->wiphy);
+
+	if (sta->mlo) {
+		clear_bit(ahsta->ml_peer_id, ah->free_ml_peer_id_map);
+		ahsta->ml_peer_id = ATH12K_MLO_PEER_ID_INVALID;
+	}
+}
+
 int ath12k_peer_mlo_link_peers_delete(struct ath12k_vif *ahvif, struct ath12k_sta *ahsta)
 {
 	struct ieee80211_sta *sta = ath12k_ahsta_to_sta(ahsta);
diff --git a/drivers/net/wireless/ath/ath12k/peer.h b/drivers/net/wireless/ath/ath12k/peer.h
index aeea5c5..cbebe91 100644
--- a/drivers/net/wireless/ath/ath12k/peer.h
+++ b/drivers/net/wireless/ath/ath12k/peer.h
@@ -24,6 +24,7 @@ int ath12k_link_sta_rhash_delete(struct ath12k_base *ab, struct ath12k_link_sta
 int ath12k_link_sta_rhash_add(struct ath12k_base *ab, struct ath12k_link_sta *arsta);
 struct ath12k_link_sta *ath12k_link_sta_find_by_addr(struct ath12k_base *ab, const u8 *addr);
 u16 ath12k_peer_ml_alloc(struct ath12k_hw *ah);
+void ath12k_peer_ml_free(struct ath12k_hw *ah, struct ath12k_sta *ahsta);
 void ath12k_mac_peer_disassoc(struct ath12k_base *ab, struct ieee80211_sta *sta,
 			      struct ath12k_sta *ahsta,
 			      enum ath12k_debug_mask debug_mask);
-- 
2.34.1

